pkgs/build-support/rust/rust-utils.nix

```diff
@@ -0,0 +1,378 @@
+# Code for buildRustCrate, a Nix function that builds Rust code, just
+# like Cargo, but using Nix instead.
+#
+# This can be useful for deploying packages with NixOps, and to share
+# binary dependencies between projects.
+
+{ lib, buildPlatform, stdenv, defaultCrateOverrides, fetchCrate, ncurses, rustc  }:
+
+let buildCrate = { crateName, crateVersion, crateAuthors, buildDependencies,
+                   dependencies, completeDeps, completeBuildDeps,
+                   crateFeatures, libName, build, release, libPath,
+                   crateType, metadata, crateBin, finalBins,
+                   verbose, colors }:
+
+      let depsDir = lib.concatStringsSep " " dependencies;
+          completeDepsDir = lib.concatStringsSep " " completeDeps;
+          completeBuildDepsDir = lib.concatStringsSep " " completeBuildDeps;
+          makeDeps = dependencies:
+            (lib.concatMapStringsSep " " (dep:
+              let extern = lib.strings.replaceStrings ["-"] ["_"] dep.libName; in
+              (if dep.crateType == "lib" then
+                 " --extern ${extern}=${dep.out}/lib/lib${extern}-${dep.metadata}.rlib"
+              else
+                 " --extern ${extern}=${dep.out}/lib/lib${extern}-${dep.metadata}${buildPlatform.extensions.sharedLibrary}")
+            ) dependencies);
+          deps = makeDeps dependencies;
+          buildDeps = makeDeps buildDependencies;
+          optLevel = if release then 3 else 0;
+          rustcOpts = (if release then "-C opt-level=3" else "-C debuginfo=2");
+          rustcMeta = "-C metadata=${metadata} -C extra-filename=-${metadata}";
+          version_ = lib.splitString "-" crateVersion;
+          versionPre = if lib.tail version_ == [] then "" else builtins.elemAt version_ 1;
+          version = lib.splitString "." (lib.head version_);
+          authors = lib.concatStringsSep ":" crateAuthors;
+      in ''
+      norm=""
```

Comment:
I think it is a bit scary to have a giant set of complex shell scripts embedded in to a .nix file. What would it take to move the shell scripts in to dedicated files, and reference them from here? Then, we can easily use static analysis to help improve / understand them later on.

Reply:
I tend to agree, but the shell script uses lots of parameters that are nix variables, and some parts are even of the form `${optionalString …}`. I guess the way to separate that script would be to create bash functions, but these are harder to read and get right than the Nix code we currently use.

Since I've never done it before, this might as well come from a big misunderstanding of what you meant, so please correct me if I'm wrong.

Reply:
I agree with @P-E-Meunier that we by not embedding shell in nix we loose useful nix features.

Reply:
`buildCrate` could become its own file though.

Reply:
I'm working on a patch to split the shell script into a separate file.  Not all of the shell code will transfer, but I think I can get most, and reduce the nix file to just initializing env vars, sourcing the shell script, and running a `start_build()` shell function or something.  At the very least, it'll be something concrete to review and compare.

Reply:
Okay.  Initial patch is drafted.  The resulting `rust-utils.nix` \[[1]()] and shell script \[[2]()] is available for viewing.

\[1]: https://nest.pijul.com/boxofrox/nix-rust:master/8ce71bef4766c2439602
\[2]: https://nest.pijul.com/boxofrox/nix-rust:master/289f899c9c9f6d041d

Reply:
@P-E-Meunier, @boxofrox's work looks great, are you keen to use it?